src/analysis/souffle/taint.dl

```diff
@@ -52,6 +52,27 @@
 // of the taint analysis).
 .decl claimNotEdge(principal: Principal, src: AccessPath, dst: AccessPath)
 
+// Indicates that an AccessPath is given an integrity tag directly. This rule
+// should only be populated with base facts with no derivation to prevent us
+// from getting tangled up in stratification issues. Propagation will happen
+// via the lacksIntegrityTag rule.
+.decl directHasIntegrityTag(accessPath: AccessPath, owner: Principal, integTag: IntegrityTag)
```

Comment:
My main issue with that is, to someone naive to the codebase, what is the difference between `hasIntegrityTag` and `mustHaveIntegrityTag`?

With confidentiality tags, there is a major difference between `hasTag` and `mayHaveTag`: the level of certainty is higher in the first than in the second, which is enough to clue a new reader into the difference in semantics. For integrity tags, `hasIntegrityTag` and `mustHaveIntegrityTag` are at the same level of certainty, and could invite confusion. I'm open to renaming this, I also think `directHasIntegrityTag` is awkward, but I haven't come up with a better human-readable way to indicate the distinction yet.

Reply:
I could have sworn I left a comment here yesterday about this.

My big problem with `hasIntegrityTag` is that the difference between `hasIntegrityTag` and `mustHaveIntegrityTag` is not clear to a human coming into the repo and trying to make sense of it. Contrast this with `hasTag` and `mayHaveTag` where there is an obvious difference in certainty.

In addition, `directHasIntegrityTag` is different from `mustHaveIntegrityTag` because `directHasIntegrityTag` is the base case, where `IntegrityTag`s are directly added, while `mustHaveIntegrityTag` includes derived facts as well.

I also am not a fan of `directHasIntegrityTag`, but I would like a name that indicates that this is just the `IntegrityTag`s that are initially added. Perhaps `baseCaseHasIntegrityTag` or `initialHasIntegrityTag`?

Reply:
On the aspect of `says_hasIntegrityTag`, I think it might be a bit premature to do that. I don't think there is functionality in the Authorization Logic parser to connect to, and I'd prefer to add that bit only when that is there (so in a follow-up PR).

Reply:
I see your point about the name. Ideally, there should be only one rule `hasIntegrityTag` or `mustHaveIntegrityTag`, but I see that can cause issues with stratification. Other choices for the name: `hasAppliedIntegrityTag`, `hasAssignedIntegrityTag`.  I don't have a strong preference as long as we document. 

As for the ownership, nothing new is needed? All we have to do is add `says_hasIntegrityTag` relation to the declaration and have the following rule just like `hasTag`:
```
hasIntegrityTag(ap, owner, tag) :- says_hasIntegrityTag(owner, ap, owner, tag).`
```

Isn't that the case?

Reply:
> Sounds good, I'll update to `hasAppliedIntegrityTag`.
> 
> On the `says_` rule, it's not that nothing new is needed, it's that I don't think the name we're connecting to exists in the parser yet. Regardless, I can add it now, not a big deal.

We don't need any support in the parser? In any case, do it in a separate PR.

Reply:
Fixed name, will defer connecting to `says_` to separate PR.